app/controllers/assignments_controller.rb

```diff
@@ -664,49 +647,25 @@ def upload_assignment_list
 
     encoding = params[:encoding]
     assignment_list = assignment_list.utf8_encode(encoding)
+    case file_format
 
-    case params[:file_format]
-      when 'csv'
-        result = MarkusCSV.parse(assignment_list) do |row|
-          assignment = Assignment.find_or_create_by(short_identifier: row[0])
-          attrs = Hash[DEFAULT_FIELDS.zip(row)]
-          attrs.delete_if { |_, v| v.nil? }
-          if assignment.new_record?
-            assignment.submission_rule = NoLateSubmissionRule.new
-            assignment.assignment_stat = AssignmentStat.new
-            assignment.token_period = 1
-            assignment.non_regenerating_tokens = false
-            assignment.unlimited_tokens = false
-          end
-          assignment.update(attrs)
-          raise CSVInvalidLineError unless assignment.valid?
-        end
-        unless result[:invalid_lines].empty?
-          flash_message(:error, result[:invalid_lines])
-        end
-        unless result[:valid_lines].empty?
-          flash_message(:success, result[:valid_lines])
-        end
-      when 'yml'
-        begin
-          map = YAML::load(assignment_list)
-          map[:assignments].map do |row|
-            row[:submission_rule] = NoLateSubmissionRule.new
-            row[:assignment_stat] = AssignmentStat.new
-            row[:token_period] = 1
-            row[:non_regenerating_tokens] = false
-            row[:unlimited_tokens] = false
-            update_assignment!(row)
-          end
-        rescue ActiveRecord::ActiveRecordError, ArgumentError => e
-          flash_message(:error, e.message)
-          redirect_to action: 'index'
-          return
+    when 'csv'
+      @result = Assignment.upload_assignment_list('csv', assignment_list, DEFAULT_FIELDS)
+      unless @result[:invalid_lines].empty?
+        flash_message(:error, @result[:invalid_lines])
+      end
+      unless @result[:valid_lines].empty?
+        flash_message(:success, @result[:valid_lines])
+      end
+
+    when 'yml'
+        @result = Assignment.upload_assignment_list('yml', assignment_list, DEFAULT_FIELDS)
+        if @result.is_a?(error)
+          flash_message(:error, @result.message)
         end
-      else
-        return
+    else
+      return
```

Comment:
Don't do the `return` here. (Redirecting to `index` is still good.)

But as in the download method, add a `flash_message` to report an incorrect format.